hbase-server/src/main/java/org/apache/hadoop/hbase/wal/WALEdit.java

```diff
@@ -52,45 +49,104 @@
  * column family name to-hand.. just shove it into the WALEdit if available. Doing this, we can
  * save on a parse of each Cell to figure column family down the line when we go to add the
  * WALEdit to the WAL file. See the hand-off in FSWALEntry Constructor.
+ * @see WALKey
  */
 // TODO: Do not expose this class to Coprocessors. It has set methods. A CP might meddle.
 @InterfaceAudience.LimitedPrivate({ HBaseInterfaceAudience.REPLICATION,
     HBaseInterfaceAudience.COPROC })
 public class WALEdit implements HeapSize {
-  private static final Logger LOG = LoggerFactory.getLogger(WALEdit.class);
-
-  // TODO: Get rid of this; see HBASE-8457
+  // Below defines are for writing WALEdit 'meta' Cells..
+  // TODO: Get rid of this system of special 'meta' Cells. See HBASE-8457. It suggests
+  // adding a type to WALEdit itself for use denoting meta Edits and their types.
   public static final byte [] METAFAMILY = Bytes.toBytes("METAFAMILY");
-  @VisibleForTesting
+
+  /**
+   * @deprecated Since 2.3.0. Not used.
+   */
+  @Deprecated
   public static final byte [] METAROW = Bytes.toBytes("METAROW");
+
+  /**
+   * @deprecated Since 2.3.0. Make it protected, internal-use only. Use
+   *   {@link #isCompactionMarker(Cell)}
+   */
+  @Deprecated
   @VisibleForTesting
   public static final byte[] COMPACTION = Bytes.toBytes("HBASE::COMPACTION");
+
+  /**
+   * @deprecated Since 2.3.0. Make it protected, internal-use only.
+   */
+  @Deprecated
   @VisibleForTesting
   public static final byte [] FLUSH = Bytes.toBytes("HBASE::FLUSH");
-  @VisibleForTesting
-  public static final byte [] REGION_EVENT = Bytes.toBytes("HBASE::REGION_EVENT");
+
+  /**
+   * Qualifier for region event meta 'Marker' WALEdits start with the
+   * {@link #REGION_EVENT_PREFIX} prefix ('HBASE::REGION_EVENT::'). After the prefix,
+   * we note the type of the event which we get from the RegionEventDescriptor protobuf
+   * instance type (A RegionEventDescriptor protobuf instance is written as the meta Marker
+   * Cell value). Adding a type suffix means we do not have to deserialize the protobuf to
+   * figure out what type of event this is.. .just read the qualifier suffix. For example,
+   * a close region event descriptor will have a qualifier of HBASE::REGION_EVENT::REGION_CLOSE.
+   * See WAL.proto and the EventType in RegionEventDescriptor protos for all possible
+   * event types.
+   */
+  private static final String REGION_EVENT_STR = "HBASE::REGION_EVENT";
+  private static final String REGION_EVENT_PREFIX_STR = REGION_EVENT_STR + "::";
+  private static final byte [] REGION_EVENT_PREFIX = Bytes.toBytes(REGION_EVENT_PREFIX_STR);
+
+  /**
+   * @deprecated Since 2.3.0. Remove. Not for external use. Not used.
+   */
+  @Deprecated
+  public static final byte [] REGION_EVENT = Bytes.toBytes(REGION_EVENT_STR);
+
+  /**
+   * We use this define figuring if we are carrying a close event.
+   */
+  private static final byte [] REGION_EVENT_CLOSE =
```

Comment:
OK so this is the trick... Seems fine?

Reply:
Ugly, but will do for now. Proper soln is typing the WALEdits themselves as suggested in cited issue. Will need a bit of work changing how we read WALS so we read a WALEdit at a time rather than Cell-at-a-time which we do in a bunch of places. etc.